File: python/iterator/iterator.py
from abc import ABC


class IIterator(ABC):
    def has_next(self) -> bool:
        pass

    def next(self):
        pass


class MenuItem:
    def __init__(self, description: str):
        self._description = description

    def get_description(self):
        return self._description


class MenuItemIterator(IIterator):
    def __init__(self, items):
        self._items = items
        self._current_position = 0

    def has_next(self) -> bool:
        return self._current_position < len(self._items)

    def next(self):
        next_item = self._items[self._current_position]
        self._current_position += 1
        return next_item


class IMenu(ABC):
    def get_iterator(self) -> IIterator:
        pass

    def add_item(self, item: MenuItem):
        pass


class Menu(IMenu):
    def __init__(self):
        self._items = []

    def get_iterator(self) -> IIterator:
        return MenuItemIterator(self._items)

    def add_item(self, item: MenuItem):
        self._items.append(item)


def main():
    menu = Menu()
    menu.add_item(MenuItem("Ice Cream"))
    menu.add_item(MenuItem("Soup"))
    it = menu.get_iterator()
    while it.has_next():
        print(it.next().get_description())


if __name__ == '__main__':
    main()

File: cpp/iterator/iterator.cpp
#include <string>
#include <utility>
#include <vector>
#include <memory>
#include <iostream>

template<typename T>
class IIterator {
public:
    virtual T next() = 0;

    virtual bool has_next() = 0;
};

class MenuItem {
public:
    explicit MenuItem(std::string description) : _description(std::move(description)) {
    }

    std::string get_description() {
        return _description;
    }

private:
    std::string _description;
};

class MenuItemIterator : public IIterator<std::shared_ptr<MenuItem>> {
public:
    explicit MenuItemIterator(const std::vector<std::shared_ptr<MenuItem>> &arr) : _arr(arr) {
        _current_pos = 0;
    }

    std::shared_ptr<MenuItem> next() override {
        return _arr[_current_pos++];
    }

    bool has_next() override {
        return _current_pos < _arr.size();
    }

private:
    const std::vector<std::shared_ptr<MenuItem>> &_arr;
    uint32_t _current_pos;
};

class IMenu {
public:
    virtual std::shared_ptr<IIterator<std::shared_ptr<MenuItem>>> get_iterator() = 0;

    virtual void add_item(std::shared_ptr<MenuItem> item) = 0;
};

class Menu : IMenu {
public:
    std::shared_ptr<IIterator<std::shared_ptr<MenuItem>>> get_iterator() override {
        return std::make_shared<MenuItemIterator>(elements);
    }

    void add_item(std::shared_ptr<MenuItem> item) override {
        elements.push_back(item);
    }

private:
    std::vector<std::shared_ptr<MenuItem>> elements;
};

int main() {
    Menu menu;
    menu.add_item(std::make_shared<MenuItem>("Ice Cream"));
    menu.add_item(std::make_shared<MenuItem>("Soup"));
    auto itr = menu.get_iterator();
    while (itr->has_next()) {
        std::cout << itr->next()->get_description() << std::endl;
    }
    return 0;
}
